napari/layers/_labels_layer/model.py

```diff
@@ -601,40 +601,35 @@ def on_mouse_release(self, event):
         """
         self._last_cursor_coord = None
 
-    def on_key_press(self, event):
-        """Called whenever key pressed in canvas.
+    def _activate_paint_mode(self):
+        self.mode = Mode.PAINT
 
-        Parameters
-        ----------
-        event : Event
-            Vispy event
-        """
-        if event.native.isAutoRepeat():
-            return
-        else:
-            if event.key == ' ':
-                if self.mode != Mode.PAN_ZOOM:
-                    self._mode_history = self.mode
-                    self.mode = Mode.PAN_ZOOM
-                else:
-                    self._mode_history = Mode.PAN_ZOOM
-            elif event.key == 'p':
-                self.mode = Mode.PAINT
-            elif event.key == 'f':
-                self.mode = Mode.FILL
-            elif event.key == 'z':
-                self.mode = Mode.PAN_ZOOM
-            elif event.key == 'l':
-                self.mode = Mode.PICKER
-
-    def on_key_release(self, event):
-        """Called whenever key released in canvas.
+    def _activate_fill_mode(self):
+        self.mode = Mode.FILL
 
-        Parameters
-        ----------
-        event : Event
-            Vispy event
-        """
-        if event.key == ' ':
-            if self._mode_history != Mode.PAN_ZOOM:
-                self.mode = self._mode_history
+    def _activate_pan_zoom_mode(self):
+        self.mode = Mode.PAN_ZOOM
+
+    def _activate_picker_mode(self):
+        self.mode = Mode.PICKER
+
+    def _hold_to_pan_zoom(self):
+        # on press
+        prev_mode = self.mode
+
+        if self.mode != Mode.PAN_ZOOM:
+            self.mode = Mode.PAN_ZOOM
+
+        yield
+
+        # on release
+        if prev_mode != Mode.PAN_ZOOM:
+            self.mode = prev_mode
+
+    default_keybindings = {
```

Comment:
I guess the reason for creating the dictionary instead of just using decorators here is so that we can reset to defaults later if we want, right?

Reply:
No, using decorators would also add it to the default keybindings anyways (as long as it is initialized as an empty dictionary). However, you cannot use the decorators in this situation because you don't have access to them – `bind_key` et al have not been defined in this scope; they are inherited from the base `Layer` class.

Reply:
ah ok so the decorators are to be used by our users / plugin developers, but internally we will use the `default_keybindings` dictionary. that's fine

Reply:
Not quite – I would say that we want to use the functional/decorator form wherever possible; it's just not possible to add these keybindings to a class within its own definition. I do know how to make it work but I'm not sure how useful it would really be.